db.go

```diff
@@ -465,51 +508,75 @@ func (db *DB) reload() (err error) {
 			continue
 		}
-		if db.beyondRetention(meta) {
-			deleteable[meta.ULID] = struct{}{}
-			continue
-		}
-		for _, b := range meta.Compaction.Parents {
-			deleteable[b.ULID] = struct{}{}
+		b, err := OpenBlock(dir, db.chunkPool)
+		if err != nil {
+			return nil, nil, err
 		}
+		goodBlocks[meta.ULID] = b
 	}
-	// Blocks we failed to open should all be those we are want to delete anyway.
-	for c, err := range corrupted {
-		if _, ok := deleteable[c]; !ok {
-			return errors.Wrapf(err, "unexpected corrupted block %s", c)
+	return goodBlocks, corruptedBlocks, nil
+}
+
+func (db *DB) filterDeleteable(goodBlocks map[ulid.ULID]*Block, corrupted map[ulid.ULID]error) (filteredBlocks []*Block, deleteable map[ulid.ULID]struct{}, err error) {
+	deleteable = make(map[ulid.ULID]struct{})
+	for _, block := range goodBlocks {
+		if db.beyondRetention(&block.meta) {
+			deleteable[block.meta.ULID] = struct{}{}
 		}
-	}
-	// Load new blocks into memory.
-	for _, dir := range dirs {
-		meta, err := readMetaFile(dir)
-		if err != nil {
-			return errors.Wrapf(err, "read meta information %s", dir)
+		for _, b := range block.meta.Compaction.Parents {
+			deleteable[b.ULID] = struct{}{}
 		}
+	}
+	var blocksSize = int64(0)
+	var blocks []*Block
+	for ulid, block := range goodBlocks {
 		// Don't load blocks that are scheduled for deletion.
-		if _, ok := deleteable[meta.ULID]; ok {
+		if _, ok := deleteable[ulid]; ok {
 			continue
 		}
-		// See if we already have the block in memory or open it otherwise.
-		b, ok := db.getBlock(meta.ULID)
-		if !ok {
-			b, err = OpenBlock(dir, db.chunkPool)
-			if err != nil {
-				return errors.Wrapf(err, "open block %s", dir)
-			}
-		}
-		blocks = append(blocks, b)
-		opened[meta.ULID] = struct{}{}
+		blocks = append(blocks, block)
+		blocksSize += block.Size()
 	}
+	// Sort the blocks to make sure the oldest is removed first.
 	sort.Slice(blocks, func(i, j int) bool {
 		return blocks[i].Meta().MinTime < blocks[j].Meta().MinTime
 	})
-	if err := validateBlockSequence(blocks); err != nil {
-		return errors.Wrap(err, "invalid block sequence")
+	// Select blocks for deletion.
+	for _, block := range blocks {
+		if db.opts.MaxBytes > 0 && blocksSize > db.opts.MaxBytes {
+			blocksSize -= block.Size()
+			deleteable[block.Meta().ULID] = struct{}{}
+			for _, b := range block.Meta().Compaction.Parents {
+				deleteable[b.ULID] = struct{}{}
+			}
+			db.metrics.dataLimitDeletions.Inc()
+		}
+	}
+	db.metrics.dataLimitDeletions.Inc()
+	// Blocks we failed to open should all be those we want to delete anyway
+	for c, err := range corrupted {
+		if _, ok := deleteable[c]; !ok {
+			return nil, nil, errors.Wrapf(err, "unexpected corrupted block %s", c)
+		}
 	}
+	// Filter out blocks that aren't scheduled for deletion
+	for ulid, block := range goodBlocks {
+		if _, ok := deleteable[ulid]; ok {
+			continue
+		}
+		filteredBlocks = append(filteredBlocks, block)
+	}
+	return filteredBlocks, deleteable, nil
```

Comment:
Those have different types though? goodBlocks is a `map[ulid.ULID]*Block`, while filteredBlocks was a `[]*Block`

Reply:
I don't think that will work though, because if you start modifying the thing that you're ranging over it's going to mess things up. For example, if index 3 and 4 are both 'deleteable', when it hits index 3, it's going to remove that value form goodBlocks, and everything after gets shifted down one index. So then it the checks index 4, but index 4 is now the original index 5, and the original index 4 is now in index 3, so it will be skipped over.

Reply:
`goodBlocks` can still be a slice, it doesn't need to be a map, but removing certain values like that will lead to issues.